File: Source/ObjC/UITableView+DropDown.m
//
//  UITableView+DropDown.m
//  DropDownTableViewObjC
//
//  Created by NSSimpleApps on 15.01.17.
//  Copyright © 2017 NSSimpleApps. All rights reserved.
//

#import "UITableView+DropDown.h"
#import "DropDownTableViewDataSource.h"
#import "DropDownTableViewDelegate.h"
#import "NSIndexPath+DropDown.h"
#import "NSArray+Functional.h"

@implementation UITableView (DropDown)

- (id<DropDownTableViewDelegate>)dropDownDelegate {
    
    return (id<DropDownTableViewDelegate>)self.delegate;
}

- (void)setDropDownDelegate:(id<DropDownTableViewDelegate>)dropDownDelegate {
    
    self.delegate = (id<UITableViewDelegate>)dropDownDelegate;
}

- (id<DropDownTableViewDataSource>)dropDownDataSource {
    
    return (id<DropDownTableViewDataSource>)self.dataSource;
}

- (void)setDropDownDataSource:(id<DropDownTableViewDataSource>)dropDownDataSource {
    
    self.dataSource = (id<UITableViewDataSource>)dropDownDataSource;
}

- (CGRect)rectForRow:(NSInteger)row {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForRow =
        [self.dropDownDataSource tableView:self indexPathsForRows:@[@(row)]].firstObject;
        
        if (indexPathForRow) {
            
            return [self rectForRowAtIndexPath:indexPathForRow];
        }
    }
    return CGRectZero;
}

// IndexPath(forSubrow: Int, inMainrow: Int)
- (CGRect)rectForSubrowAtIndexPath:(NSIndexPath *)indexPath {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForSubrow =
        [self.dropDownDataSource tableView:self indexPathsForSubrows:@[@(indexPath.subrow)] inRow:indexPath.mainrow].firstObject;
        
        if (indexPathForSubrow) {
            
            return [self rectForRowAtIndexPath:indexPathForSubrow];
        }
    }
    return CGRectZero;
}

- (nullable id<DropDownIndexPath>)dropDownIndexPathAtPoint:(CGPoint)point {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForRow = [self indexPathForRowAtPoint:point];
        
        if (indexPathForRow) {
            
            return
            [self.dropDownDataSource valueForIndexPath:indexPathForRow
                                           valueForRow:^id<DropDownIndexPath> _Nullable(NSInteger row) {
                                               
                                               return @(row);
                                           }
                                        valueForSubrow:^id<DropDownIndexPath> _Nullable(NSInteger subrow, NSInteger row) {
                                            
                                            return [NSIndexPath indexPathForSubrow:subrow inMainrow:row];
                                        }];
        }
    }
    return nil;
}

- (nullable id<DropDownIndexPath>)dropDownIndexPathForCell:(UITableViewCell *)cell {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPath = [self indexPathForCell:cell];
        
        if (indexPath) {
            
            return
            [self.dropDownDataSource valueForIndexPath:indexPath
                                           valueForRow:^id<DropDownIndexPath> _Nullable(NSInteger row) {
                                               
                                               return @(row);
                                           }
                                        valueForSubrow:^id<DropDownIndexPath> _Nullable(NSInteger subrow, NSInteger row) {
                                            
                                            return [NSIndexPath indexPathForSubrow:subrow inMainrow:row];
                                        }];
        }
    }
    return nil;
}

- (nullable NSArray<NSNumber *> *)rowsInRect:(CGRect)rect {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPaths = [self indexPathsForRowsInRect:rect];
        
        if (indexPaths) {
            
            return
            [indexPaths nsk_flatMapUsingBlock:^id _Nullable(NSIndexPath * _Nonnull indexPath) {
                
                return
                [self.dropDownDataSource valueForIndexPath:indexPath
                                               valueForRow:^NSNumber * _Nullable(NSInteger row) {
                                                   
                                                   return @(row);
                                               }
                                            valueForSubrow:^NSNumber * _Nullable(NSInteger subrow, NSInteger row) {
                                                
                                                return nil;
                                            }];
            }];
        }
    }
    return nil;
}

- (nullable NSArray<NSIndexPath *> *)indexPathsForSubrowsInRect:(CGRect)rect {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPaths = [self indexPathsForRowsInRect:rect];
        
        if (indexPaths) {
            
            return
            [indexPaths nsk_flatMapUsingBlock:^id _Nullable(NSIndexPath * _Nonnull indexPath) {
                
                return
                [self.dropDownDataSource valueForIndexPath:indexPath
                                               valueForRow:^NSIndexPath * _Nullable(NSInteger row) {
                                                   
                                                   return nil;
                                               }
                                            valueForSubrow:^NSIndexPath * _Nullable(NSInteger subrow, NSInteger row) {
                                                
                                                return [NSIndexPath indexPathForSubrow:subrow inMainrow:row];
                                            }];
            }];
        }
    }
    return nil;
}

- (nullable __kindof UITableViewCell *)cellForRowAtRow:(NSInteger)row {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForRow =
        [self.dropDownDataSource tableView:self indexPathsForRows:@[@(row)]].firstObject;
        
        if (indexPathForRow) {
            
            return [self cellForRowAtIndexPath:indexPathForRow];
        }
    }
    return nil;
}

// IndexPath(forSubrow: Int, inMainrow: Int)
- (nullable __kindof UITableViewCell *)cellForSubrowAtIndexPath:(NSIndexPath *)indexPath {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForSubrow =
        [self.dropDownDataSource tableView:self indexPathsForSubrows:@[@(indexPath.subrow)] inRow:indexPath.mainrow].firstObject;
        
        if (indexPathForSubrow) {
            
            return [self cellForRowAtIndexPath:indexPathForSubrow];
        }
    }
    return nil;
}

- (nullable NSArray<NSNumber *> *)visibleRows {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForVisibleRows = self.indexPathsForVisibleRows;
        
        if (indexPathsForVisibleRows) {
            
            return
            [indexPathsForVisibleRows nsk_flatMapUsingBlock:^id _Nullable(NSIndexPath * _Nonnull indexPath) {
                
                return
                [self.dropDownDataSource valueForIndexPath:indexPath
                                               valueForRow:^NSNumber * _Nullable(NSInteger row) {
                                                   
                                                   return @(row);
                                               }
                                            valueForSubrow:^NSNumber * _Nullable(NSInteger subrow, NSInteger row) {
                                                
                                                return nil;
                                            }];
            }];
        }
    }
    return nil;
}

// [IndexPath(forSubrow: Int, inMainrow: Int)]
- (nullable NSArray<NSIndexPath *> *)indexPathsForVisibleSubrows {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForVisibleRows = self.indexPathsForVisibleRows;
        
        if (indexPathsForVisibleRows) {
            
            return
            [indexPathsForVisibleRows nsk_flatMapUsingBlock:^id _Nullable(NSIndexPath * _Nonnull indexPath) {
                
                return
                [self.dropDownDataSource valueForIndexPath:indexPath
                                               valueForRow:^NSIndexPath * _Nullable(NSInteger row) {
                                                   
                                                   return nil;
                                               }
                                            valueForSubrow:^NSIndexPath * _Nullable(NSInteger subrow, NSInteger row) {
                                                
                                                return [NSIndexPath indexPathForSubrow:subrow inMainrow:row];
                                            }];
            }];
        }
    }
    return nil;
}

- (void)scrollToRowAtRow:(NSInteger)row atScrollPosition:(UITableViewScrollPosition)scrollPosition animated:(BOOL)animated {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForRow =
        [self.dropDownDataSource tableView:self indexPathsForRows:@[@(row)]].firstObject;
        
        if (indexPathForRow) {
            
            [self scrollToRowAtIndexPath:indexPathForRow atScrollPosition:scrollPosition animated:animated];
        }
    }
}

// IndexPath(forSubrow: Int, inMainrow: Int)
- (void)scrollToSubrowAtIndexPath:(NSIndexPath *)indexPath atScrollPosition:(UITableViewScrollPosition)scrollPosition animated:(BOOL)animated {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForSubrow =
        [self.dropDownDataSource tableView:self indexPathsForSubrows:@[@(indexPath.subrow)] inRow:indexPath.mainrow].firstObject;
        
        if (indexPathForSubrow) {
            
            [self scrollToRowAtIndexPath:indexPathForSubrow atScrollPosition:scrollPosition animated:animated];
        }
    }
}

- (void)insertRowsAtRows:(NSArray<NSNumber *> *)rows withRowAnimation:(UITableViewRowAnimation)animation {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForRows =
        [self.dropDownDataSource tableView:self indexPathsForRows:rows];
        
        if (indexPathsForRows) {
            
            [self.dropDownDataSource dropDownInsertRows:rows inTableView:self];
            [self insertRowsAtIndexPaths:indexPathsForRows withRowAnimation:animation];
        }
    }
}

- (void)insertSubrows:(NSArray<NSNumber *> *)subrows inRow:(NSInteger)row withRowAnimation:(UITableViewRowAnimation)animation {
    
    NSArray<NSIndexPath *> *indexPaths =
    [subrows nsk_mapUsingBlock:^NSIndexPath * _Nonnull(NSNumber * _Nonnull subrow) {
        
        return [NSIndexPath indexPathForRow:row + subrow.integerValue + 1 inSection:0];
    }];
    
    [self insertRowsAtIndexPaths:indexPaths withRowAnimation:animation];
}

- (void)deleteRowsAtRows:(NSArray<NSNumber *> *)rows withRowAnimation:(UITableViewRowAnimation)animation {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForRows =
        [self.dropDownDataSource tableView:self indexPathsForRows:rows];
        
        if (indexPathsForRows) {
            
            [self.dropDownDataSource dropDownDeleteRows:rows inTableView:self];
            [self deleteRowsAtIndexPaths:indexPathsForRows withRowAnimation:animation];
        }
    }
}

- (void)deleteSubrows:(NSArray<NSNumber *> *)subrows inRow:(NSInteger)row withRowAnimation:(UITableViewRowAnimation)animation {
    
    NSArray<NSIndexPath *> *indexPaths =
    [subrows nsk_mapUsingBlock:^NSIndexPath * _Nonnull(NSNumber * _Nonnull subrow) {
        
        return [NSIndexPath indexPathForRow:row + subrow.integerValue + 1 inSection:0];
    }];
    
    [self deleteRowsAtIndexPaths:indexPaths withRowAnimation:animation];
}

- (void)reloadRowsAtRows:(NSArray<NSNumber *> *)rows withRowAnimation:(UITableViewRowAnimation)animation {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForRows =
        [self.dropDownDataSource tableView:self indexPathsForRows:rows];
        
        if (indexPathsForRows) {
            
            [self reloadRowsAtIndexPaths:indexPathsForRows withRowAnimation:animation];
        }
    }
}

- (void)reloadSubrows:(NSArray<NSNumber *> *)subrows inRow:(NSInteger)row withRowAnimation:(UITableViewRowAnimation)animation {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForSubrows =
        [self.dropDownDataSource tableView:self indexPathsForSubrows:subrows inRow:row];
        
        if (indexPathsForSubrows) {
            
            [self reloadRowsAtIndexPaths:indexPathsForSubrows withRowAnimation:animation];
        }
    }
}

- (void)moveRowAtRow:(NSInteger)row toRow:(NSInteger)newRow {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForRow =
        [self.dropDownDataSource tableView:self indexPathsForRows:@[@(row)]].firstObject;
        
        NSIndexPath *indexPathForNewRow =
        [self.dropDownDataSource tableView:self indexPathsForRows:@[@(newRow)]].firstObject;
        
        if (indexPathForRow && indexPathForNewRow) {
            
            [self moveRowAtIndexPath:indexPathForRow toIndexPath:indexPathForNewRow];
        }
    }
}

- (void)moveSubrow:(NSInteger)subrow toNewSubrow:(NSInteger)newSubrow inRow:(NSInteger)row {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForSubrows =
        [self.dropDownDataSource tableView:self indexPathsForSubrows:@[@(subrow), @(newSubrow)] inRow:row];
        
        NSIndexPath *indexPathForSubrow = indexPathsForSubrows.firstObject;
        NSIndexPath *indexPathForNewSubrow = indexPathsForSubrows.lastObject;
        
        if (indexPathForSubrow && indexPathForNewSubrow) {
            
            [self moveRowAtIndexPath:indexPathForSubrow toIndexPath:indexPathForNewSubrow];
        }
    }
}

/// returns Int for row or IndexPath(forSubrow: Int, inMainrow: Int) for subrow
- (nullable id<DropDownIndexPath>)dropDownIndexPathForSelectedRow {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForSelectedRow = self.indexPathForSelectedRow;
        
        if (indexPathForSelectedRow) {
            
            return
            [self.dropDownDataSource valueForIndexPath:indexPathForSelectedRow
                                           valueForRow:^id<DropDownIndexPath> _Nullable(NSInteger row) {
                                               
                                               return @(row);
                                           }
                                        valueForSubrow:^id<DropDownIndexPath> _Nullable(NSInteger subrow, NSInteger row) {
                                            
                                            return [NSIndexPath indexPathForSubrow:subrow inMainrow:row];
                                        }];
        }
    }
    return nil;
}

- (nullable NSArray<NSNumber *> *)rowsForSelectedRows {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForSelectedRows = self.indexPathsForSelectedRows;
        
        if (indexPathsForSelectedRows) {
            
            return
            [indexPathsForSelectedRows nsk_flatMapUsingBlock:^id _Nullable(NSIndexPath * _Nonnull indexPath) {
                
                return
                [self.dropDownDataSource valueForIndexPath:indexPath
                                               valueForRow:^NSNumber * _Nullable(NSInteger row) {
                                                   
                                                   return @(row);
                                               }
                                            valueForSubrow:^NSNumber * _Nullable(NSInteger subrow, NSInteger row) {
                                                
                                                return nil;
                                            }];
            }];
        }
    }
    return nil;
}

// IndexPath(forSubrow: Int, inMainrow: Int)
- (nullable NSArray<NSIndexPath *> *)indexPathsForSelectedSubrows {
    
    if (self.dropDownDataSource) {
        
        NSArray<NSIndexPath *> *indexPathsForSelectedRows = self.indexPathsForSelectedRows;
        
        if (indexPathsForSelectedRows) {
            
            return
            [indexPathsForSelectedRows nsk_flatMapUsingBlock:^id _Nullable(NSIndexPath * _Nonnull indexPath) {
                
                return
                [self.dropDownDataSource valueForIndexPath:indexPath
                                               valueForRow:^NSIndexPath * _Nullable(NSInteger row) {
                                                   
                                                   return nil;
                                               }
                                            valueForSubrow:^NSIndexPath * _Nullable(NSInteger subrow, NSInteger row) {
                                                
                                                return [NSIndexPath indexPathForSubrow:subrow inMainrow:row];
                                            }];
            }];
        }
    }
    return nil;
}

- (void)selectRow:(nullable NSNumber *)row animated:(BOOL)animated scrollPosition:(UITableViewScrollPosition)scrollPosition {
    
    if (row != nil && self.dropDownDataSource != nil) {
        
        NSIndexPath *indexPathForRow =
        [self.dropDownDataSource tableView:self indexPathsForRows:@[row]].firstObject;
        
        if (indexPathForRow) {
            
            [self selectRowAtIndexPath:indexPathForRow animated:animated scrollPosition:scrollPosition];
        }
        
    } else {
        
        [self selectRowAtIndexPath:nil animated:animated scrollPosition:scrollPosition];
    }
}

// IndexPath(forSubrow: Int, inMainrow: Int)
- (void)selectSubrowAtIndexPath:(nullable NSIndexPath *)indexPath animated:(BOOL)animated scrollPosition:(UITableViewScrollPosition)scrollPosition {
    
    if (indexPath != nil && self.dropDownDataSource != nil) {
        
        NSIndexPath *indexPathForSubrow =
        [self.dropDownDataSource tableView:self indexPathsForSubrows:@[@(indexPath.subrow)] inRow:indexPath.mainrow].firstObject;
        
        if (indexPathForSubrow) {
            
            [self selectRowAtIndexPath:indexPathForSubrow animated:animated scrollPosition:scrollPosition];
        }
        
    } else {
        
        [self selectRowAtIndexPath:nil animated:animated scrollPosition:scrollPosition];
    }
}

- (void)deselectRow:(NSInteger)row animated:(BOOL)animated {
    
    if (self.dropDownDataSource != nil) {
        
        NSIndexPath *indexPathForRow =
        [self.dropDownDataSource tableView:self indexPathsForRows:@[@(row)]].firstObject;
        
        if (indexPathForRow) {
            
            [self deselectRowAtIndexPath:indexPathForRow animated:animated];
        }
    }
}

// IndexPath(forSubrow: Int, inMainrow: Int)
- (void)deselectSubrowAtIndexPath:(NSIndexPath *)indexPath animated:(BOOL)animated {
    
    if (self.dropDownDataSource) {
        
        NSIndexPath *indexPathForSubrow =
        [self.dropDownDataSource tableView:self indexPathsForSubrows:@[@(indexPath.subrow)] inRow:indexPath.mainrow].firstObject;
        
        if (indexPathForSubrow) {
            
            [self deselectRowAtIndexPath:indexPathForSubrow animated:animated];
        }
    }
}

@end

File: Source/Swift/UITableView+DropDown.swift
//
//  UITableView+DropDown.swift
//  DropDownTableView
//
//  Created by NSSimpleApps on 03.04.16.
//  Copyright © 2016 NSSimpleApps. All rights reserved.
//

import UIKit

public protocol DropDownIndexPath { }

extension Int: DropDownIndexPath { }
extension IndexPath: DropDownIndexPath { }

public extension UITableView {
    
    public var dropDownDelegate: DropDownTableViewDelegate? {
        
        get {
            
            return self.delegate as? DropDownTableViewDelegate
        }
        
        set {
            
            self.delegate = newValue as? UITableViewDelegate
        }
    }
    
    public var dropDownDataSource: DropDownTableViewDataSource? {
        
        get {
            
            return self.dataSource as? DropDownTableViewDataSource
        }
        
        set {
            
            return self.dataSource = newValue as? UITableViewDataSource
        }
    }
    
    public func rectForRow(_ row: Int) -> CGRect {
        
        if let indexPathForRow = self.dropDownDataSource?.tableView(self, indexPathsForRows: [row]).first {
            
            return self.rectForRow(at: indexPathForRow)
        }
        
        return CGRect.zero
    }
    
    // IndexPath(forSubrow: Int, inMainrow: Int)
    public func rectForSubrow(at indexPath: IndexPath) -> CGRect {
        
        if let indexPathForSubrow = self.dropDownDataSource?.tableView(self, indexPathsForSubrows: [indexPath.subrow], inRow: indexPath.mainrow).first {
            
            return self.rectForRow(at: indexPathForSubrow)
        }
        return CGRect.zero
    }
    
    /// returns Int for row or IndexPath(forSubrow: Int, inMainrow: Int) for subrow
    public func dropDownIndexPath(at point: CGPoint) -> DropDownIndexPath? {
        
        if let indexPath = self.indexPathForRow(at: point) {
            
            return self.dropDownDataSource?.valueForIndexPath(indexPath,
                                                              valueForRow: { (row: Int) -> DropDownIndexPath in
                                                                
                                                                return row
            
            }, valueForSubrow: { (subrow, row) -> DropDownIndexPath in
                    
                    return IndexPath(forSubrow: subrow, inMainrow: row)
            })
        }
        return nil
    }
    
    /// returns Int for row or IndexPath(forSubrow: Int, inMainrow: Int) for subrow
    public func dropDownIndexPath(for cell: UITableViewCell) -> DropDownIndexPath? {
        
        if let indexPath = self.indexPath(for: cell) {
            
            return self.dropDownDataSource?.valueForIndexPath(indexPath,
                                                              valueForRow: { (row: Int) -> DropDownIndexPath in
                                                                
                                                                return row
            },
                                                              valueForSubrow: { (subrow, row) -> DropDownIndexPath in
                                                                
                                                                return IndexPath(forSubrow: subrow, inMainrow: row)
            })
        }
        return nil
    }
    
    public func rows(in rect: CGRect) -> [Int]? {
        
        if let indexPaths = self.indexPathsForRows(in: rect) {
            
            return indexPaths.flatMap({ (indexPath: IndexPath) -> Int? in
                
                self.dropDownDataSource?.valueForIndexPath(indexPath,
                    valueForRow: { (row) -> Int in
                        
                        return row
                    },
                    valueForSubrow: { (subrow, row) -> Int? in
                        
                        return nil
                })
            })
        }
        return nil
    }
    
    // IndexPath(forSubrow: Int, inMainrow: Int)
    public func indexPathsForSubrows(in rect: CGRect) -> [IndexPath]? {
        
        if let indexPaths = self.indexPathsForRows(in: rect) {
            
            return indexPaths.flatMap({ (indexPath: IndexPath) -> IndexPath? in
                
                self.dropDownDataSource?.valueForIndexPath(indexPath,
                    valueForRow: { (row) -> IndexPath? in
                        
                        return nil
                    },
                    valueForSubrow: { (subrow, row) -> IndexPath in
                        
                        return IndexPath(forSubrow: subrow, inMainrow: row)
                })
            })
        }
        return nil
    }
    
    public func cellForRow(at row: Int) -> UITableViewCell? {
        
        if let indexPathForRow = self.dropDownDataSource?.tableView(self, indexPathsForRows: [row]).first {
            
            return self.cellForRow(at: indexPathForRow)
        }
        return nil
    }
    
    // IndexPath(forSubrow: Int, inMainrow: Int)
    public func cellForSubrow(at indexPath: IndexPath) -> UITableViewCell? {
        
        if let indexPathForSubrow = self.dropDownDataSource?.tableView(self, indexPathsForSubrows: [indexPath.subrow], inRow: indexPath.mainrow).first {
            
            return self.cellForRow(at: indexPathForSubrow)
        }
        return nil
    }
    
    public var visibleRows: [Int]? {
        
        return self.indexPathsForVisibleRows?.flatMap({ (indexPath: IndexPath) -> Int? in
            
            self.dropDownDataSource?.valueForIndexPath(indexPath,
                                                       valueForRow: { (row: Int) -> Int in
                                                        
                                                        return row
            },
                                                       valueForSubrow: { (subrow, row) -> Int? in
                                                        
                                                        return nil
            })
        })
    }
    
    // IndexPath(forSubrow: Int, inMainrow: Int)
    public var indexPathsForVisibleSubrows: [IndexPath]? {
        
        return self.indexPathsForVisibleRows?.flatMap({ (indexPath: IndexPath) -> IndexPath? in
                
            self.dropDownDataSource?.valueForIndexPath(indexPath,
                    valueForRow: { (row: Int) -> IndexPath? in
                        
                        return nil
                        
                    },
                    valueForSubrow: { (subrow, row) -> IndexPath in
                        
                        return IndexPath(forSubrow: subrow, inMainrow: row)
                })
            })
    }
    
    public func scrollToRow(at row: Int, at scrollPosition: UITableViewScrollPosition, animated: Bool) {
        
        if let indexPathForRow = self.dropDownDataSource?.tableView(self, indexPathsForRows: [row]).first {
            
            self.scrollToRow(at: indexPathForRow, at: scrollPosition, animated: animated)
        }
    }
    
    // IndexPath(forSubrow: Int, inMainrow: Int)
    public func scrollToSubrow(at indexPath: IndexPath, at scrollPosition: UITableViewScrollPosition, animated: Bool) {
        
        if let indexPathForSubrow = self.dropDownDataSource?.tableView(self, indexPathsForSubrows: [indexPath.subrow], inRow: indexPath.mainrow).first {
            
            self.scrollToRow(at: indexPathForSubrow, at: scrollPosition, animated: animated)
        }
    }
    
    public func insertRows(at rows: [Int], with animation: UITableViewRowAnimation) {
        
        if let indexPathsForRows = self.dropDownDataSource?.tableView(self, indexPathsForRows: rows) {
        
            self.dropDownDataSource?.dropDownInsertRows(rows, in: self)
            
            self.insertRows(at: indexPathsForRows, with: animation)
        }
    }
    
    public func insertSubrows(_ subrows: [Int], in row: Int, with animation: UITableViewRowAnimation) {
        
        let indexPaths = subrows.map { (subrow: Int) -> IndexPath in
            
            IndexPath(row: row + subrow + 1, section: 0)
        }
        
        self.insertRows(at: indexPaths, with: animation)
    }
    
    public func deleteRows(at rows: [Int], with animation: UITableViewRowAnimation) {
        
        if let indexPathsForRows = self.dropDownDataSource?.tableView(self, indexPathsForRows: rows) {
            
            self.dropDownDataSource?.dropDownDeleteRows(rows, in: self)
            
            self.deleteRows(at: indexPathsForRows, with: animation)
        }
    }
    
    public func deleteSubrows(_ subrows: [Int], in row: Int, with animation: UITableViewRowAnimation) {
        
        let indexPaths = subrows.map { (subrow: Int) -> IndexPath in
            
            IndexPath(row: row + subrow + 1, section: 0)
        }
        
        self.deleteRows(at: indexPaths, with: animation)
    }
    
    public func reloadRows(at rows: [Int], with animation: UITableViewRowAnimation) {
        
        if let indexPathsForRows = self.dropDownDataSource?.tableView(self, indexPathsForRows: rows) {
            
            self.reloadRows(at: indexPathsForRows, with: animation)
        }
    }
    
    public func reloadSubrows(_ subrows: [Int], in row: Int, with animation: UITableViewRowAnimation) {
        
        if let indexPathsForSubrows = self.dropDownDataSource?.tableView(self, indexPathsForSubrows: subrows, inRow: row) {
                
            self.reloadRows(at: indexPathsForSubrows, with: animation)
        }
    }
    
    public func moveRow(at row: Int, to newRow: Int) {
        
        if let indexPathForRow = self.dropDownDataSource?.tableView(self, indexPathsForRows: [row]).first, let indexPathForNewRow = self.dropDownDataSource?.tableView(self, indexPathsForRows: [newRow]).first {
            
            self.moveRow(at: indexPathForRow, to: indexPathForNewRow)
        }
    }
    
    public func moveSubrow(_ subrow: Int, to newSubrow: Int, in row: Int) {
        
        let indexPathsForSubrows = self.dropDownDataSource?.tableView(self, indexPathsForSubrows: [subrow, newSubrow], inRow: row)
        
        if let indexPathForSubrow = indexPathsForSubrows?.first, let indexPathForNewSubrow = indexPathsForSubrows?.last {
            
            self.moveRow(at: indexPathForSubrow, to: indexPathForNewSubrow)
        }
    }
    
    /// returns Int for row or IndexPath(forSubrow: Int, inMainrow: Int) for subrow
    public var dropDownIndexPathForSelectedRow: DropDownIndexPath? {
        
        if let indexPathForSelectedRow = self.indexPathForSelectedRow {
            
            return self.dropDownDataSource?.valueForIndexPath(indexPathForSelectedRow,
                                                              valueForRow: { (row) -> DropDownIndexPath in
                                                                
                                                                return row
            },
                                                              valueForSubrow: { (subrow, row) -> DropDownIndexPath in
                                                                
                                                                return IndexPath(forSubrow: subrow, inMainrow: row)
            })
        }
        
        return nil
    }
    
    public var rowsForSelectedRows: [Int]? {
            
        return self.indexPathsForSelectedRows?.flatMap({ (indexPath: IndexPath) -> Int? in
            
            self.dropDownDataSource?.valueForIndexPath(indexPath,
                                                       valueForRow: { (row) -> Int in
                                                            
                                                            return row
                                                            
                },
                                                       valueForSubrow: { (subrow, row) -> Int? in
                                                            
                                                            return nil
            })
        })
    }
    
    // IndexPath(forSubrow: Int, inMainrow: Int)
    public var indexPathsForSelectedSubrows: [IndexPath]? {
            
        return self.indexPathsForSelectedRows?.flatMap({ (indexPath: IndexPath) -> IndexPath? in
            
            self.dropDownDataSource?.valueForIndexPath(indexPath,
                                                          valueForRow: { (row) -> IndexPath? in
                                                            
                                                            return nil
                                                            
                },
                                                          valueForSubrow: { (subrow, row) -> IndexPath in
                                                            
                                                            return IndexPath(forSubrow: subrow, inMainrow: row)
            })
        })
    }
    
    public func select(row: Int?, animated: Bool, scrollPosition: UITableViewScrollPosition) {
        
        if let row = row, let indexPathForRow = self.dropDownDataSource?.tableView(self, indexPathsForRows: [row]).first {
            
            self.selectRow(at: indexPathForRow, animated: animated, scrollPosition: scrollPosition)
        
        } else {
            
            self.selectRow(at: nil, animated: animated, scrollPosition: scrollPosition)
        }
    }
    
    // IndexPath(forSubrow: Int, inMainrow: Int)
    public func selectSubrow(at indexPath: IndexPath?, animated: Bool, scrollPosition: UITableViewScrollPosition) {
        
        if let indexPath = indexPath, let indexPathForSubrow = self.dropDownDataSource?.tableView(self, indexPathsForSubrows: [indexPath.subrow], inRow: indexPath.mainrow).first {
                
                self.selectRow(at: indexPathForSubrow, animated: animated, scrollPosition: scrollPosition)
            
        } else {
            
            self.selectRow(at: nil, animated: animated, scrollPosition: scrollPosition)
        }
    }
    
    public func deselect(row: Int, animated: Bool) {
        
        if let indexPathForRow = self.dropDownDataSource?.tableView(self, indexPathsForRows: [row]).first {
            
            self.deselectRow(at: indexPathForRow, animated: animated)
        }
    }
    
    // IndexPath(forSubrow: Int, inMainrow: Int)
    public func deselectSubrow(at indexPath: IndexPath, animated: Bool) {
        
        if let indexPathForSubrow = self.dropDownDataSource?.tableView(self, indexPathsForSubrows: [indexPath.subrow], inRow: indexPath.mainrow).first {
            
            self.deselectRow(at: indexPathForSubrow, animated: animated)
        }
    }
}
